### PR TITLE
feat(webview): thin telemetry middleware + generic TelemetryRunner (@microsoft/vscode-ext-webview 0.10.0)

### PR DESCRIPTION
**Pull request by tnaum-ms**
## Summary

De-engineers the `@microsoft/vscode-ext-webview` telemetry middleware and adopts the new shape across the extension's webviews. Based on feedback from the `vscode-cosmosdb` integration attempt, the telemetry body becomes a **thin delegator** and the `TelemetryRunner` becomes **generic over the context enrichment**, so a consumer's runner owns the telemetry scope, chooses what to contribute to `ctx`, and classifies the outcome. Duration and error classification are no longer duplicated by the package — the runner's backend (e.g. `callWithTelemetryAndErrorHandling`) already records them.

Bumps the package `0.9.1` → `0.10.0` (minor, not patch — see **Breaking changes**).

## Package changes (`@microsoft/vscode-ext-webview`)

- **Thin telemetry body + generic runner** (breaking): `TelemetryRunner<TEnrichment>.run(eventId, invocation, invoke)`; `telemetryMiddlewareBody(runner, { buildEventId })` is curried and wires as `publicProcedure.use(telemetryMiddlewareBody(runner, options))`; the body no longer stamps duration/result.
- **`ProcedureLogger`** gains symmetric optional `onStart` / `onEnd` hooks (replaces the single `log`). Both fire at the **dispatch layer** (`attachTrpc`), paired one-to-one: `onStart` runs once before each query/mutation/subscription, and a matching `onEnd` always follows — including on early setup failure. An `onStart`-only logger passed through the panel option now receives events.
- **`mergeRouters`** is now returned from `initWebviewTrpc()` and re-exported from the shared entry (no more direct `@trpc/server` import to compose routers).
- **`AnyRouter`** is re-exported from `.` and `./react` (was already on `./host`).

See `packages/vscode-ext-webview/MIGRATION.md` for the full `0.9.1 → 0.10.0` upgrade guide, and the updated `ADVANCED.md` / `README.md`.

## Extension changes

- The DocumentDB telemetry runner injects the full `IActionContext` as `ctx.actionContext`. `actionContext` is **not** a field on the base `RouterContext`; it is modeled as an additive enrichment (`WithTelemetry<T> = T & RpcEnrichment` in `_integration/trpc.ts`). Instrumented procedures (`publicProcedureWithTelemetry`) narrow to `WithTelemetry<RouterContext>` to read `ctx.actionContext.telemetry`; plain `publicProcedure` procedures narrow to bare `RouterContext`, so reading `actionContext` there is a **compile error** instead of a runtime `undefined`. Controllers build the plain `RouterContext` (no `Omit<...>` root type).
- Query Insights composes its main + events routers via the exposed `mergeRouters` (retires the flat-record workaround).
- `rpcConcurrencyLogger` adopts the `onEnd` hook.
- The `.github/skills/` authoring guides (`webview-trpc-messaging`, `telemetry-instrumentation`) and the stale `trpcToTelemetry` code comments in `queryInsightsEventsRouter.ts` are updated to the new `telemetryMiddlewareBody` / `TelemetryRunner` / `ctx.actionContext` surface.

## Breaking changes

The package is pre-1.0; the telemetry/logger API changes are breaking but internal — only this extension and `vscode-cosmosdb` consume the package. The release is a **minor** bump (`0.10.0`), not a patch: a patch (`0.9.2`) satisfies the common `^0.9.1` range (`>=0.9.1 <0.10.0`), so consumers would have received the breaking release on a clean install. The minor bump moves it outside that range, making the upgrade opt-in. All changes are documented in `MIGRATION.md`.

## Review dispositions

- **Copilot — context loss on `next({ ctx })`:** not reproducible. tRPC v11 shallow-merges the override over the existing context (`{ ...opts.ctx, ...nextOpts.ctx }`), so base fields (`signal`, `sessionId`, `clusterId`, …) are preserved. Thread resolved.
- **Copilot — unsound `getInfo` cast:** fixed. `documentsView.getInfo` is now instrumented, and `actionContext` moved off the base context, so the cast is honest and `tsc` enforces the invariant everywhere.

## Verification

- `npm run build` ✅ · `npm run lint` ✅
- repo `jest`: **2674 passing**
- package `jest`: **95 passing**

> Note: this branch intentionally excludes the `indexView` webview migration (out of scope — part of #732, not yet on `main`).

### CODE REVIEW

**Comment by tnaum-ms** [comment]
### F1 addressed — breaking changes now ship as `0.10.0` (not `0.9.2`)

**Commit:** 0f3ed21e2a47e09f21a5292cd9c352554c86e887

**What changed**
- `packages/vscode-ext-webview/package.json`: `0.9.2` → `0.10.0`
- `MIGRATION.md`: heading is now `0.9.1 → 0.10.0`, the *After* code label matches, and a short paragraph explains the minor-bump rationale
- `package-lock.json`: workspace entry synced to `0.10.0`

**Why**
This PR intentionally makes source-breaking changes (generic `TelemetryRunner`, curried `telemetryMiddlewareBody`, removed `WithTelemetry`, `ProcedureLogger.log` → `onStart?`/`onEnd?`). For a pre-1.0 package, npm still treats a patch as caret-compatible: `0.9.2` satisfies the common `^0.9.1` range (`>=0.9.1 <0.10.0`). Consumers on that range would therefore receive the breaking release on a clean install and fail to compile without touching their manifest.

Bumping the **minor** to `0.10.0` moves the release outside `^0.9.1`, so upgrades become opt-in. This keeps the "pre-1.0 may break on minor" convention the migration guide already documents, without the silent-patch hazard.

**Comment by tnaum-ms** [comment]
### F2 addressed — `ProcedureLogger.onStart` now fires at the dispatch layer

**Commit:** 76fbf1d95f75a3dae9445db9a1d673b825f0cd46

**What changed** (`packages/vscode-ext-webview/src/host/attachTrpc.ts`)
- Added a `logProcedureStart` helper that mirrors the existing `logProcedure` (`onEnd`) helper.
- `handleDefaultMessage` (queries/mutations) and `handleSubscriptionMessage` (subscriptions) now call `logProcedureStart({ type, path })` once, before the procedure runs.
- Placed the start call before each handler's `try`, so an early setup failure still produces a start plus its matching `onEnd` with `ok: false` (one-to-one pairing).
- Updated the `attachTrpc` `logger` JSDoc.
- `attachTrpc.test.ts`: added transport tests for `onStart`/`onEnd` ordering, exactly-once delivery, early subscription-setup failure, and an `onStart`-only logger.
- `ADVANCED.md`: documented the dispatch-layer `onStart`/`onEnd` behavior and pairing guarantees.

**Why**
`ProcedureLogger` advertises symmetric optional `onStart` / `onEnd` hooks, and the middleware body (`loggingMiddlewareBody`) already fired both — but the dispatch path used by `attachTrpc` / `WebviewController` / `openWebview` only fired `onEnd`. A consumer passing an `onStart`-only logger through the documented panel telemetry option received nothing, contradicting both the interface and the PR summary's "symmetric hooks" claim.

Both hooks remain optional (the common case is still `onEnd`-only). Start entries carry only `type` and `path`, since no duration or outcome exists yet. All 19 `attachTrpc` tests pass.

**Comment by tnaum-ms** [comment]
### F4 addressed — skill guides updated to the `ctx.actionContext` surface

**Commit:** 34dc09ee24b35836cce8dfd3ef08cd46f37ffd20

**What changed**
- `.github/skills/webview-trpc-messaging/SKILL.md`
  - Removed the `type WithTelemetry` import and every `ctx as WithTelemetry<RouterContext>` cast; procedures now narrow with `ctx as RouterContext` and read `ctx.actionContext.telemetry`.
  - Key-files table: `trpc.ts` row no longer lists `WithTelemetry` (now the `RpcEnrichment` shape contributed to `ctx.actionContext`).
  - Replaced the `trpcToTelemetry (file-local)` narrative with `telemetryMiddlewareBody` delegating to the DocumentDB `TelemetryRunner`; fixed the `publicProcedure` vs `publicProcedureWithTelemetry` comparison table, the abort/subscription examples, and the pitfalls list.
  - Frontmatter description: `trpcToTelemetry` → `telemetryMiddlewareBody`.
- `.github/skills/telemetry-instrumentation/SKILL.md`
  - The "tRPC Procedures (auto-telemetry)" example now reads `ctx.actionContext.telemetry` after narrowing to `RouterContext`.
  - Left the `callWithTelemetryAndErrorHandling(...)` examples that use `ctx.telemetry` unchanged — in those the callback `ctx` **is** the `IActionContext`, so `ctx.telemetry` is correct there.

**Why**
These are the canonical authoring references the repo instructions direct contributors (and Copilot) to follow. This PR deleted `WithTelemetry` and moved telemetry from a fixed `ctx.telemetry` slot to an injected `ctx.actionContext`, so the guides were teaching code that no longer compiles or uses the wrong context shape. They're now aligned with the shipped routers (e.g. `queryInsightsRouter.ts` uses `myCtx.actionContext.telemetry`).

**Comment by tnaum-ms** [comment]
### F5 addressed — refreshed stale `trpcToTelemetry` comments

**Commit:** 797f007bd282dad77e273b22e9259c4f27d02110

**What changed** (`src/webviews/documentdb/collectionView/queryInsights/queryInsightsEventsRouter.ts`)
- Three explanatory comments (the `STAGE3_COMPLETION_EVENT` doc, the `CompletionTelemetry` accumulator doc, and the inline note in the subscription handler) referenced `trpcToTelemetry` wrapping `opts.next()` and recording onto `ctx.telemetry`.
- Updated them to the current model: the telemetry middleware (`telemetryMiddlewareBody`) wraps the invocation — which for a subscription resolves at generator-creation time — and the deleted buffered procedure's keys mirror onto `ctx.actionContext.telemetry`.

**Why**
`trpcToTelemetry` no longer exists after this PR's reshape, so the comments misdescribed how the dedicated Stage 3 completion event relates to middleware timing. Comment-only change; no behavior change.

---

_All independent findings (F1, F2, F4, F5) are now addressed. F3/C2 (the `getInfo` context cast) is a Copilot-thread item — I'll handle it in that review thread once its fix is confirmed._

File: src/webviews/documentdb/documentView/documentsViewRouter.ts [hunk 1]
```diff
@@ -49,7 +58,7 @@ export const documentsViewRouter = router({
         .input(z.string())
         // procedure type
         .query(async ({ input, ctx }) => {
-            const myCtx = ctx as WithTelemetry<RouterContext>;
+            const myCtx = ctx as RouterContext;
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply src/webviews/documentdb/documentView/documentsViewRouter.ts:52]: Fixed in efe086ee98532ec7dba1789b5c57f01b242f386d.

You're right that the `ctx as RouterContext` assertion was unsound: `getInfo` ran on plain `publicProcedure`, so the telemetry runner never injected `actionContext`, yet the cast claimed it was present. Harmless today (the body only `JSON.stringify`s the context) but a latent `TypeError` for the first `myCtx.actionContext` access.

Rather than a local `Omit<RouterContext, 'actionContext'>`, we inverted the model so the type reflects reality:

- `actionContext` is no longer a field on the base `RouterContext`. It's modeled as an additive enrichment: `WithTelemetry<T> = T & RpcEnrichment` (new, in `trpc.ts`).
- Instrumented (`publicProcedureWithTelemetry`) procedures that read telemetry narrow to `WithTelemetry<RouterContext>`; plain `publicProcedure` procedures narrow to bare `RouterContext`, so reading `actionContext` there is now a **compile error** instead of a runtime `undefined`.
- Controllers build the plain `RouterContext` (the `Omit<...>` root type is gone).
- `documentsView.getInfo` is now instrumented (`publicProcedureWithTelemetry`) to match its `collectionView` sibling, so its cast is honest.

`tsc` flagged every `actionContext` read on a bare `RouterContext`; all were reconciled. Full build + 319 webview tests pass. The `webview-trpc-messaging` / `telemetry-instrumentation` skill guides were updated to teach the `WithTelemetry<RouterContext>` narrowing.

File: packages/vscode-ext-webview/src/host/middleware/telemetry.ts [hunk 2]
```diff
@@ -6,148 +6,143 @@
...
-            if (result.error?.message) {
-                telemetry.properties.errorMessage = result.error.message;
-            }
-        }
+        let captured: TResult | undefined;
+        await runner.run(eventId, invocation, async (enrichment) => {
+            const result = await invocation.next({ ctx: enrichment });
+            captured = result;
+            return result;
+        });
```
[the remark this answers is not in the document]
Comment (tnaum-ms) [reply packages/vscode-ext-webview/src/host/middleware/telemetry.ts:142]: No change needed here — this isn't reproducible against the installed tRPC.

`telemetryMiddlewareBody` calls `invocation.next({ ctx: enrichment })`, but tRPC v11 does **not** replace the context on a `next({ ctx })` override — it shallow-merges the override on top of the existing context. In `@trpc/server@11.10.0` the middleware dispatcher recurses with:

```js
ctx: nextOpts?.ctx ? { ...opts.ctx, ...nextOpts.ctx } : opts.ctx
```

(compiled at `node_modules/@trpc/server/dist/initTRPC-RoZMIBeA.mjs:263`, from the `unstable-core-do-not-import` procedure builder). So the enrichment is layered over the base context: `signal`, `sessionId`, `clusterId`, etc. are preserved unless the enrichment reuses a key — and it only adds `actionContext`. This is tRPC's standard middleware context-extension behavior.

Consistent with that, the enriched procedures in this PR read base fields (`myCtx.clusterId`, `myCtx.signal`, `myCtx.sessionId`, …) right alongside `myCtx.actionContext`, and the full build + 319 webview tests pass with no `undefined` context fields.

Leaving `telemetryMiddlewareBody` as-is. (Happy to add a regression test that starts from a non-empty base context and asserts both base and enrichment fields survive, to lock this against a future tRPC bump — say the word.)

**Comment by tnaum-ms** [comment]
### F3 addressed — `actionContext` is now an additive `WithTelemetry` enrichment

**Commit:** efe086ee98532ec7dba1789b5c57f01b242f386d
(also answered inline in the Copilot C2 thread: https://github.com/microsoft/vscode-documentdb/pull/795#discussion_r3586028299)

**What changed**
- `trpc.ts`: added `export type WithTelemetry<T> = T & RpcEnrichment` and refreshed the `RpcEnrichment` docs.
- `appRouter.ts`: re-exports `WithTelemetry`.
- `collectionViewRouter.ts` / `documentsViewRouter.ts`: removed `actionContext` from `RouterContext`; the instrumented procedures that read it now narrow to `WithTelemetry<RouterContext>`.
- `queryInsights/queryInsightsRouter.ts`: Stage 1 / Stage 2 casts narrowed to `WithTelemetry<RouterContext>`.
- `collectionViewController.ts` / `documentsViewController.ts`: root context is now plain `RouterContext` (the `Omit<RouterContext, 'actionContext'>` is gone).
- **D1:** `documentsView.getInfo` is now `publicProcedureWithTelemetry` (+ `WithTelemetry<RouterContext>`), matching its `collectionView` sibling — so its cast is honest.
- Updated the `webview-trpc-messaging` and `telemetry-instrumentation` skill guides to teach the `WithTelemetry<RouterContext>` narrowing.

**Why**
`RouterContext` declared `actionContext` as required, but the telemetry runner only injects it for `publicProcedureWithTelemetry`. The uninstrumented `getInfo` cast `ctx as RouterContext` therefore asserted a field that wasn't there — harmless today (it only `JSON.stringify`s ctx) but a latent `TypeError` for the first `myCtx.actionContext` access. Inverting the model (actionContext out of the base type, added back via `WithTelemetry<T>`) makes reading `actionContext` on a plain `publicProcedure` a **compile error** instead of a runtime `undefined`. `tsc` flagged all 28 `actionContext` reads on bare `RouterContext`; each was reconciled. Full build + 319 webview tests pass.

---

_Bookkeeping:_ commit 5f07c5f5 is a formatting-only follow-up to F4 (Prettier table-column re-alignment in `webview-trpc-messaging/SKILL.md`) — no content change.

**Comment by tnaum-ms** [comment]
### Final documentation & 0.10.0 migration review

**Commit:** 87f37666b202552c3f76103fc4e99ea756d92dd9

Swept all manuals, docs, and skills for `0.10.0` consistency and stale telemetry references. Two shipped-doc fixes:

- **`README.md`** — Status line `0.9.0-preview` → `0.10.0` (now matches `package.json`).
- **`MIGRATION.md`** — added the additive `WithTelemetry<T> = T & { actionContext }` consumer pattern as the recommended alternative to baking `actionContext` into `RouterContext` + `Omit<…>`, so an uninstrumented (plain `publicProcedure`) procedure can't assert an `actionContext` that was never injected. This is the pattern the reference extension adopted in F3.

**Reviewed and confirmed already-consistent (no change needed):**
- `MIGRATION.md` heading/labels are `0.9.1 → 0.10.0`; the `0.9.2` mentions that remain are the intentional "why a minor, not a patch" rationale.
- `ADVANCED.md` documents the dispatch-layer `onStart`/`onEnd` (F2) and the `telemetryMiddlewareBody` + `TelemetryRunner` flow.
- `README.md` / `src/host/middleware/README.md` reflect the current `ProcedureLogger` (`onStart?`/`onEnd?`) surface.
- Both skill guides (`webview-trpc-messaging`, `telemetry-instrumentation`) teach `WithTelemetry<RouterContext>`; the remaining `ctx as RouterContext` examples are signal-only and correct.
- Repo `docs/user-manual` and `docs/release-notes` contain no package/telemetry references. Remaining `trpcToTelemetry`/`0.9.2` hits are confined to historical `docs/ai-and-plans/PRs/**` planning archives (accurate as history) and `CONTRIBUTING.md` generic version examples.

**Verification (full PR checklist):**
- `npm run l10n` — no bundle changes (no user-facing strings added/changed/removed)
- Prettier — clean
- `npm run lint` — clean
- `npx jest --no-coverage` — **2674 passed / 159 suites**
- `npm run build` — clean (package builds as `@microsoft/vscode-ext-webview@0.10.0`)

**Comment by tnaum-ms** [comment]
### Follow-up — skill example consistency

**Commit:** 4c7cf5c94682b7fe95dacf51d03c7c8440a0b6dd

Audited every `ctx as WithTelemetry<RouterContext>` occurrence after the F3 change. All casts in shipped code and current docs are correct (build + 2674 tests confirm they compile). The `docs/ai-and-plans/PRs/**` hits are frozen planning archives (historical), and `MIGRATION.md`'s `0.9.1`-labelled one is the intentional "before" example of the removed package-level helper.

One internal inconsistency fixed: the `webview-trpc-messaging` section-2 `getData` example cast to `WithTelemetry<RouterContext>` but never read `actionContext`, which contradicted the least-privilege rule elsewhere in the same guide ("narrow with `WithTelemetry` **when the procedure reads telemetry**, `RouterContext` otherwise"). The example now writes a telemetry property, so the cast is justified and the guide is self-consistent.

**Comment by tnaum-ms** [comment]
@microsoft-github-policy-service

**Comment by tnaum-ms** [comment]
approved ✅

**Review by bk201- — APPROVED** [review]
